### PR TITLE
coord: only import "valid" indexes into dataflows

### PR DESCRIPTION
**Pull request by sploiselle**
### Motivation

This PR fixes a recognized bug. #8021

### Description

The `DataflowBuilder` must consider indexes for which dataflows have not yet been built as "invalid." When importing a view into a dataflow that had an invalid index, nothing prevented the `DataflowBuilder` from importing the view's invalid indexes.

### Checklist

- [x] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

**Review by benesch — APPROVED** [review]
So yeah, I'm super sketched out by the fact that we're importing all these indexes here, but I think this is the right fix for now. LGTM modulo a test!

**Comment by benesch** [comment]
BTW, rather than "invalid", I'd say that it's "indexes for which dataflows have not yet been created." The indexes have to create dataflows in _some_ topologically valid order. The confusion stems from the fact that the catalog knows about all of them at boot, but the dataflow layer only knows about what it's been told, and the code that constructs each dataflow needs to be careful to construct that dataflow only out of already existing dataflows, not dataflows that will exist in the future. If that makes any sense.

All this code is probably due for some clarifying refactors, but again I think this is a great fix for now. TYVM for tracking it down!

**Comment by sploiselle** [comment]
Ah, I used the term "invalid indexes" only because indexes that do not cause this bug to trip are referred to as "[valid indexes](https://github.com/MaterializeInc/materialize/blob/c481021578fecf3577de96798dce260ced2e01c2/src/coord/src/coord/dataflow_builder.rs#L48-L49)," and was trying to draw a comparison. Appreciate the clarification.

**Comment by benesch** [comment]
Ah yeah, I think that terminology has the scope of that one local variable there, vs. the `dataflow` crate. I think it's a case of rewriting the sentence

> The dataflow layer is considers some indexes "invalid," and isn't aware of them.

to

> The `DataflowBuilder` must consider indexes for which dataflows have not yet been built as "invalid."

**Comment by sploiselle** [comment]
Sorry for the confusion I caused there

File: test/regression/github-8021.td [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# by the Apache License, Version 2.0.
+
+#
+# Ensure Materialize can restart with this dependency graph
+#
+
+> CREATE TABLE t1 (f1 int4, f2 int4);
+> CREATE VIEW v2 AS SELECT * FROM t1;
+> CREATE INDEX i1 ON v2 (f1, f2);
+> CREATE INDEX i2 ON t1 (f2);
```
Comment (benesch) [inline test/regression/github-8021.td:17]: I think this should work just as a normal testdrive test? In CI it runs with `—validate-catalog` which essentially simulates a catalog restart! Mind giving that a whirl and seeing if I’m right?

Comment (benesch) [reply test/regression/github-8021.td:17]: Hrm, actually that probably doesn’t test the actual ship dataflow part. Maybe get @philip-stoev’s opinion tomorrow am on whether there’s an existing place to do these restart tests?

**Review by mjibson — APPROVED** [review]

**Comment by sploiselle** [comment]
@benesch Refactored the `mzworkflow` bit. This has the advantage of making it easier  to ensure these tests get run in CI, but all of these disparate tests sharing a catalog is a little weird. Is this akin to what you had in mind or have I sufficiently bungled the implementation?

**Comment by sploiselle** [comment]
@benesch says via Slack:

> I think merge it for now!
> But would be good to get the experimental mode tests in the same framework

Will move over other tests in a subsequent PR, as well as add a release note.

**Pull request merged**